howto/android/custom-vhal.md

```diff
@@ -136,6 +136,100 @@ addons:
+           </hal>
+       </manifest>
+
+   Place the VHAL manifest fragment file in the same folder as the `Android.bp` file that declares the VHAL service, and include the Anbox VHAL manifest fragment in the VHAL service declaration within the `Android.bp` file. Additionally, add the HIDL module as a shared library that the VHAL service links to in the same file.
+
+
+       cc_binary {
+           name: "vendor.<company>.vehicle@<version>-service",
+           vintf_fragments: [
+               ...
+               "vendor.canonical.interfaces.vehicle@1.0.xml",
+           ],
+           shared_libs: [
+               ...
+               "vendor.canonical.interfaces.vehicle@1.0",
+           ],
+           ...
+       }
+
+1. Take the [example](https://github.com/canonical/vendor_canonical_interfaces/tree/main/vehicle/1.0/default) as a reference, which implements the `IVehicle` interface.
+
+        Return<void> VHalService::get(
+            const VehiclePropValue& requestedPropValue, IVehicle::get_cb _hidl_cb) {
+          uid_t uid = android::IPCThreadState::self()->getCallingUid();
+          if (uid != AID_VEHICLE_NETWORK) {
+            _hidl_cb(StatusCode::ACCESS_DENIED, kEmptyValue);
+            return Void();
+          }
+
+          // NOTE: a VHAL implementation must allow access to non-readable vehicle properties.
+          return Void();
+        }
+
+        Return<StatusCode> VHalService::set(const VehiclePropValue& value) {
+          uid_t uid = android::IPCThreadState::self()->getCallingUid();
+          if (uid != AID_VEHICLE_NETWORK)
+            return StatusCode::ACCESS_DENIED;
+
+          // NOTE: a VHAL implementation must allow modification of non-writable vehicle properties.
+          return StatusCode::NOT_AVAILABLE;
+        }
+
+   Please note that each function must implement a security mechanism to restrict access to vehicle properties to the authorized `AID_VEHICLE_NETWORK` process.
+
+1. Instantiate the VHAL service that implements the interface and register it as a binder service in the VHAL implementation.
+
+        #include <VHalService.h>
+
+        int main(int /* argc */, char* /* argv */[]) {
+            ...
+            ...
+            configureRpcThreadpool(4, true);
+            auto vendor_vhal_service = std::make_unique<VHalService>();
+            status_t status = vendor_vhal_service->registerAsService();
+            if (status != OK) {
+                return 1;
+            }
+            joinRpcThreadpool();
+            return 1;
+        }
+
+After implementing the Anbox HIDL interfaces and integrating them into your VHAL implementation, [build](https://source.android.com/docs/setup/build/building) the VHAL module. Then follow the instructions for [customising the Anbox VHAL](https://documentation.ubuntu.com/anbox-cloud/en/latest/howto/android/custom-vhal/) and load it as an addon during the Android runtime. Once registered, the service can be accessed by the Anbox VHAL adapter.
```

Comment:
This reads like integrating this interface is the first step and then comes customising Anbox VHAL. But this topic has the instructions in reverse (this is the same custom VHAL topic that is being referred circularly).

Would it make sense to have this integration first in the topic and then the steps about customising VHAL?

Reply:
That's a really good suggestion. I understand where you're coming from. From my perspective, the integration of the Anbox HIDL interface is a step beyond the first section (customizing VHAL, as I’ve been reluctant to call it), and it requires more low-level knowledge and AOSP development, rather than just following instructions and side-loading a third-party VHAL implementation through an addon to make it work with the Anbox runtime. It's more like an advanced feature for average users.

On the other hand, if this chapter is intended as developer-oriented technical documentation, then yes, we should definitely place the section on integrating the Anbox VHAL at the beginning of this page.

For now, the 'customizing Anbox VHAL' section is more about loading a (third-party) VHAL implementation instead of the default one and making it work with the Anbox runtime. Strictly speaking, it doesn't customize the Anbox VHAL. That's something we might want to consider using a better term for.

Reply:
It is more "custom VHAL implementation" than "customizing the Anbox VHAL". So "custom implementation" is something the user will use instead of the default we provide. It does not mean the user is customizing the Anbox VHAL but rather replaces with their own preferred VHAL implementation. I will make sure it reflects the "custom VHAL implementation" idea everywhere in this topic.